src/photos/ducks/clustering/albums.js

```diff
@@ -173,34 +180,73 @@ const photoExistsInCollection = (photos, photo) => {
 
  * @param {Object[]} newPhotos - Set of new photos to clusterize
  * @param {Object[]} albums - Set of existing auto albums
- * @returns {Object} dict associating albums to a set of photos.
+ * @returns {Map} Map associating a set of albums to a set of photos.
  *
  */
 export const albumsToClusterize = async (newPhotos, albums) => {
-  const clusterize = {}
-
-  for (const newPhoto of newPhotos) {
-    // Find clusters matching this newPhoto
-    const matchingAlbums = getMatchingClusters(newPhoto, albums)
-    if (matchingAlbums.length > 0) {
-      const key = matchingAlbums[1]
-        ? matchingAlbums[0]._id + ':' + matchingAlbums[1]._id
-        : matchingAlbums[0]._id
-      if (key in clusterize) {
-        if (!photoExistsInCollection(clusterize[key], newPhoto)) {
-          clusterize[key].push(newPhoto)
+  const clusterize = new Map()
+  try {
+    for (const newPhoto of newPhotos) {
+      // Find clusters matching the photo
+      const matchingAlbums = getMatchingClusters(newPhoto, albums)
+      if (matchingAlbums.length > 0) {
+        let keyExists = false
+
+        for (const key of clusterize.keys()) {
+          if (difference(matchingAlbums, key).length < 1) {
+            // The matchingAlbums are included in the key: add the photo
+            clusterize.set(key, insertNewPhoto(clusterize.get(key), newPhoto))
+            keyExists = true
+            break
+          } else if (intersection(matchingAlbums, key).length > 1) {
+            // The matchingAlbums partially exist into the key: merge it
+            const mergedKey = union(key, matchingAlbums)
+            let mergedValues = clusterize.get(key)
+            const missingAlbums = difference(matchingAlbums, key)
+
+            for (const album of missingAlbums) {
+              if (clusterize.has(album)) {
+                // Album exists as a key in the Map: remove it from the Map and
+                // save its photos
+                mergedValues = mergedValues.concat(clusterize.get(album))
+                clusterize.delete(album)
+              } else {
+                // Album does not exist in the Map: add it with its photos
+                const photosToClusterize = await findPhotosToReclusterize([
+                  album
+                ])
+                mergedValues = mergedValues.concat(photosToClusterize)
+              }
+            }
+            mergedValues.push(newPhoto)
+            // resort values
+            const sortedMergedValues = mergedValues.sort(
+              (a, b) => a.timestamp - b.timestamp
```

Comment:
There are 3 different places where we re-sort the album, do we really need to do it that often?

Reply:
Well, there are indeed 3 cases where we sort: 
1. When we add a new photo to a sorted set of photos. [We resort the whole array](https://github.com/cozy/cozy-drive/blob/829e957643d245a8195573c0db8a425d959aff07/src/photos/ducks/clustering/albums.js#L170), as I assumed it was smart enough to not recompute the whole array (I might be wrong)
2. When we query photos referenced by albums. [We sort all the photos](https://github.com/cozy/cozy-drive/blob/829e957643d245a8195573c0db8a425d959aff07/src/photos/ducks/clustering/albums.js#L159) after querying each album separately as we cannot group it into a single sorted couchdb query.
3. When we "merge" albums. We need to [re-sort the set of photos](https://github.com/cozy/cozy-drive/blob/829e957643d245a8195573c0db8a425d959aff07/src/photos/ducks/clustering/albums.js#L224) into a single sorted set. 

The case 1. could be replaced by a function if the sort is not "smart". The case 2. should be expressed by a single query, but it is not possible from what I know (maybe in cozy-client?). The case 3. seems mandatory.